### DISCUSSION TITLE
Pass set-cookie header from fetch inside getServerSideProps

### DISCUSSION THREAD

**Discussion by vilasp** [vercel/next.js#33979] [Help] [answered]
Hi! 

I'm new to NextJS and I have a question about `getServerSideProps`.

I would like to,
1. From within `getServerSideProps` make a `fetch` request to my backend (rails).
2. Get the `Set-cookie` header in the response from that `fetch` request 
3. Add the `Set-cookie` header to the `ctx.response` object i get from the `getServerSideProps` props. 

I have the following code,
```javascript
export const getServerSideProps: GetServerSideProps = async ({ res }) => {
    const fetchResponse = await fetch('someUrl',{method:post, credentials: 'same-origin'})
    res.setHeader('set-cookie', fetchResponse.headers.get('set-cookie'))
}
```

But I always get `null` back from`fetchResponse.headers.get('set-cookie')`. 
![image](https://user-images.githubusercontent.com/5450557/152511935-a4ba47de-34c1-4e5d-bfc3-260244ddec89.png)

I have tried the `fetch` request from code in a `useEffect` inside a NextJS page and that returns the correct header and the browser sets the cookie.

Is this at all possible to do?

**What I've read**
Browser implementations of [fetch](https://fetch.spec.whatwg.org/#forbidden-response-header-name) does not allow reading the `Set-cookie` header from client side code. But from what I can see NextJS uses a [polyfill](https://github.com/vercel/next.js/blob/canary/packages/next/server/node-polyfill-fetch.js) which builds on `node-fetch`. According to their [tests](https://github.com/node-fetch/node-fetch/blob/44c67b11a36f83d2c9e0cf1671043f2a87c058f9/test/test.js#L1212-L1231) they have support to read this header.

**Comment by macsikora** [comment, marked as answer]
Hey, follow up this question. I have the same problem. I want to set browser cookies but the api call is done via next server side. Cookies set by API server do not end in the browser so this is a dead end. The problem here is that any stateful end points which set such things are no use for nextjs.

What I would want is a way to pass such thing to the browser. Any possible solutions?

**Reply by icyJoseph** [reply]
Hi,

```tsx
import type { GetServerSideProps, NextPage } from "next";

const Home: NextPage = () => {
  return <div>Hello World</div>;
};

export const getServerSideProps: GetServerSideProps = async (ctx) => {
  const response = await fetch("http://localhost:3001/cookie", {
    method: "POST",
    credentials: "same-origin",
  });

  const cookie = response.headers.get("set-cookie");

  if (!cookie) return { notFound: true };

  ctx.res.setHeader("set-cookie", cookie);

  return { props: {} };
};

export default Home;
```

I have this setup and I see the cookie coming down to the client. Could it be something with the architecture? Are there more layers between the server and the client? 

The server on `:3001`, is an express server with code I found on Stack Overflow, it just sets a cookie with cookie parser, and I see it from Next.js.  The cookie I set has this options:

```js
 let options = {
    maxAge: 1000 * 60 * 15, // would expire after 15 minutes
    httpOnly: true, // The cookie only accessible by the web server
    signed: true, // Indicates if the cookie should be signed
  };

  // Set cookie
  res.cookie("cookieName", "cookieValue", options); // options is optional
```

**Reply by macsikora** [reply]
Tnx @icyJoseph, yes just exactly came to the same thing:
```
    ctx.res.setHeader('set-cookie', apiResp.headers['set-cookie']);
```
Though I was not able to find this in any docs. I think this is quite common scenario. For me contr-intuitive was that I return things for the React (props) and I need to mutate resp object. I would think headers should be also in return... but yes it is solved. Tnx

**Reply by icyJoseph** [reply]
Still curious about @vilasp situation, hopefully this helps out!

**Reply by vilasp** [reply]
> Still curious about @vilasp situation, hopefully this helps out!

Great to hear and see that it works your way at least :) I will do some further testing when I get home to see if it could be a infrastructural problem.

**Reply by vilasp** [reply]
Did some testing over the weekend and definitely something is happening outside of Next and fetch responses for me. Tried in a isolated environment and a cookie is passed to Next from backend. 

I don't know where the issue lies but will get back with an update when I've figured it out.

**Reply by vilasp** [reply]
I will mark this as the answer since it shows how it should work. In my case thought it was an issue with understanding the underlying mechanism of how `Rails` works. It turned out that the call I was making from Next did get through the `Rails` request pipeline which showed that the cookie was attached to the `set-cookie` header in the response of the controller. It was however later in a middleware stripped because it did not passed the `CSRF` check which was in place for `Rails`.

Lesson learned 📖🎓: Check your middlewares/response pipelines and don't be lazy if you don't understand how to properly debug the libs you are using (looking at you `Rails`), the truth is out there 👽 .

**Reply by zargold** [reply]
Note... using:
```
response.headers.get("set-cookie");
```

will result in getting mangled cookies.

If your server responds potentially with multiple cookies you will need to use:

```
response.headers.raw()['set-cookie']
```
which outputs cookies as an array of strings.

**Reply by firedev99** [reply]
if your server handles mutiple cookies at the same time, this might help you out 🍻
```
export const getServerSideProps: GetServerSideProps = async (ctx) => {
  const res = await fetch(`${process.env.NEXT_PUBLIC_URL}/api/xyz`)
  const cookie = res.headers.getSetCookie()
  ctx.res.setHeader("Set-Cookie", cookie)

  return {
    props: {},
  }
}
```

**Comment by sambeetpanda507** [comment]
```js
export const getServerSideProps: GetServerSideProps = async (context) => {
  const parsedCookie = parseCookies(context) //this method comes from a package called 'nookies'
  console.log(parsedCookie)

  const accessToken = parsedCookie['refresh_token']

  if (!accessToken) {
    return {
      props: {
        auth: false,
        user: null,
      },
    }
  }

  try {
    const res = await _axios.get('get-access-token', {
      withCredentials: true,
      headers: {
        Cookie: `refresh_token=${accessToken};`,
      },
    })
    return {
      props: {
        auth: res.data.auth,
        user: res.data.userData,
      },
    }
  } catch (e) {
    return {
      props: {
        auth: false,
        user: null,
      },
    }
  }
}
```

**Reply by sambeetpanda507** [reply]
I tried to send the cookie in this way to the server and it worked for me.

**Reply by JonasSchweigler** [reply]
Thx, that's what I searched for 🥳

**Comment by 521xueweihan** [comment]
My solutions：

SET/GET cookie fail, It cloud be not set request with cookie:

```javascript
  const data: RequestInit = {};
  data.credentials = 'include';
  data.headers = {
    'Content-Type': 'application/json',
    Cookie: cookie,
  };
  data.method = 'POST';
  data.body = JSON.stringify({ code: code, state: state });

  const result: User = await fetcher('/', data);
  return result;
```

And cookie domain has **same domain name** like: .github.com

**Comment by tangye1234** [comment]
Because nextjs polyfills the Response in node side with node-fetch, so I always use this function to extract all headers from Web Response.

```typescript
export function extractHeaderKeysFromWebResponseToServerResponse(
  response: Response,
  res: ServerResponse,
  ...headerKeys: string[]
) {
  const keys = headerKeys.length ? headerKeys : [...response.headers.keys()]
  if ('getAll' in response.headers) {
    // get real all headers from node fetch headers
    keys.forEach(key => {
      const values = (response.headers as any).getAll(key) as string[]
      const cb = key.match(/^(x-|content-|via|cache-)/i) ? setHeader : appendHeader
      values.forEach(value => value && cb(res, key, value))
    })
  } else {
    keys.forEach(key => {
      const value = response.headers.get(key)
      const cb = key.match(/^(x-|content-|via|cache-)/i) ? setHeader : appendHeader
      value && cb(res, key, value)
    })
  }
}
```

Then you can extract set-cookie like this
```typescript
extractHeaderKeysFromWebResponseToServerResponse(fetchResponse, ctxRes, 'set-cookie')
```
That's all!

**Reply by tangye1234** [reply]
More information about set-cookie in web response object:
set-cookie is a forbidden header(which cannot be rewrite or modified) in Response headers.
see [https://developer.mozilla.org/en-US/docs/Glossary/Forbidden_response_header_name](https://developer.mozilla.org/en-US/docs/Glossary/Forbidden_response_header_name)
